File: languages/ruby/lib/benchmarkers/impl/quick_sort_benchmarker.rb
# Implements QuickSort benchmarking logic
require 'json'

require_relative '../benchmarker'
require_relative '../../operations/quick_sort'

class QuickSortBenchmarker < Benchmarker
  def initialize()
    super("QuickSort")
  end

  def consume_file_input(input_file_path)
    json_raw = File.read(input_file_path)
    @unsorted_list = JSON.parse(json_raw)
  end

  def get_operation_execution_time()
    arr_copy = @unsorted_list.map(&:clone)
    super() { QuickSortOperation.quick_sort(arr_copy) }
  end
end

File: languages/c#/src/benchmarkers/impl/QuickSortBenchmarker.cs
namespace Benchmarkers
{
    using System.Diagnostics;
    using System.IO;
    using System.Text.Json;

    using static Operations.QuickSortOperation;

    public class QuickSortBenchmarker : Benchmarker
    {
        private string[] unsortedList;

        public QuickSortBenchmarker() : base("QuickSort")
        {
            unsortedList = [];
        }

        public override void ConsumeInputFile(string inputFilePath)
        {
            string jsonData = File.ReadAllText(inputFilePath);
            unsortedList = JsonSerializer.Deserialize<string[]>(jsonData)!;
        }

        public override double GetOperationExecutionTime()
        {
            string[] arrCopy = (string[]) unsortedList.Clone();

            Stopwatch stopWatch = new();
            stopWatch.Start();
            QuickSort(arrCopy);
            stopWatch.Stop();

            return stopWatch.Elapsed.TotalNanoseconds;
        }
    }
}